File: aurora-core/src/main/java/net/gree/aurora/domain/cluster/ClusterRepository.java
package net.gree.aurora.domain.cluster;

import org.sisioh.dddbase.lifecycle.sync.SyncEntityIterableReader;
import org.sisioh.dddbase.lifecycle.sync.SyncRepository;

import java.util.List;

/**
 * {@link Cluster}のためのリポジトリ。
 */
public interface ClusterRepository
        extends SyncRepository<ClusterRepository, ClusterId, Cluster>,
        SyncEntityIterableReader<ClusterId, Cluster> {

    /**
     * エンティティをリストとして取得する。
     *
     * @return {@link Cluster}のリスト
     */
    List<Cluster> toList();

}

File: aurora-scala/src/main/scala/net/gree/aurora/scala/domain/cluster/ClusterRepository.scala
package net.gree.aurora.scala.domain.cluster

import net.gree.aurora.domain.cluster.ClusterRepositoryFactory
import net.gree.aurora.domain.cluster.{ClusterRepository => JClusterRepository}
import org.sisioh.dddbase.core.lifecycle.sync.SyncRepository
import scala.language.implicitConversions

/**
 * [[net.gree.aurora.domain.cluster.ClusterRepository]]のScala版。
 */
trait ClusterRepository extends SyncRepository[ClusterId, Cluster]

/**
 * コンパニオンオブジェクト。
 */
object ClusterRepository {

  /**
   * JavaオブジェクトからScalaオブジェクトを生成する。
   *
   * @param underlying [[net.gree.aurora.domain.cluster.ClusterRepository]]
   * @return [[net.gree.aurora.scala.domain.cluster.ClusterRepository]]
   */
  private[scala] def apply(underlying: JClusterRepository): ClusterRepository =
    new ClusterRepositoryImpl(underlying)

  /**
   * ファクトリメソッド。
   *
   * @return [[net.gree.aurora.scala.domain.cluster.ClusterRepository]]
   */
  def apply(): ClusterRepository =
    apply(ClusterRepositoryFactory.create())

  /**
   * ScalaオブジェクトからJavaオブジェクトに変換する。
   *
   * @param self [[net.gree.aurora.scala.domain.cluster.ClusterRepository]]
   * @return [[net.gree.aurora.domain.cluster.ClusterRepository]]
   */
  implicit def toJava(self: ClusterRepository) = self match {
    case s: ClusterRepositoryImpl => s.underlying
  }

}
